iceberg: remove manifest entry distinct_counts

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
The manifest entry schema used by Redpanda to deserialize manifests
(defined in manifest_entry_type.cc[1]) does not include distinct_counts, as
the field has been removed from the Iceberg spec, but we were previously
trying to still deserialize the distinct_counts field in
manifest_entry_values.cc[2].

This meant that we were deserializing a distinct_counts field with
values of what should have been nan_value_counts. And that we were
deserializing a nan_value_counts fields with what should have been
lower_bounds.

Because lower_bounds (map<int, binary>) and nan_value_counts (map<int,
long>) have different types, this meant that when a manifest included
entries with a non-null lower_bounds fields, Redpanda would fail to
deserialize, returning the following error:

```
ERROR 2025-08-15 13:20:44,446 [shard 1:main] iceberg - metadata_io.h:166 - Exception while parsing iceberg::manifest (path: "redpanda/topic/metadata/52d25526-4b0a-4c2e-b7bf-0c93506d53ae-m4.avro"): std::invalid_argument (Value is not the expected type)
```

or with the recently updated logging:

```
Error parsing 'nan_value_counts' map: Value of 'val' is not the expected type long: actual binary(size_bytes=1)
```

This commit fixes the deserialization error and updates the test
manifest to include a lower and upper bound field that would have caught
this error.

Longer term we should really de/serialize all fields (rather than just
the required fields), and try to figure out a way to make manifest
serialization less error prone. That is left as future work though,
given this bug on its own is a small fix.

[1] https://github.com/redpanda-data/redpanda/blob/dc2a0537ff85dcf048d01eaddfce830ee78c4971/src/v/iceberg/manifest_entry_type.cc#L58-L71
[2] https://github.com/redpanda-data/redpanda/blob/dc2a0537ff85dcf048d01eaddfce830ee78c4971/src/v/iceberg/manifest_entry_values.cc#L219


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [ ] v24.3.x

## Release Notes

### Bug Fixes

* Fixes a bug in Redpanda Iceberg manifest deserialization that would cause Redpanda to fail to commit new Iceberg files following an external operation that wrote manifests with the `lower_bounds` field.
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
iceberg: add name to primitive types

Adds ::name() to the primitive values so that they may be used by a
subsequent commit to include the names in a templatized function.

andrwng committed:
iceberg: give more information about bad type

Without more information, it's difficult to debug what fails in manifest
entry serialization. This adds a bit more context when there's an issue
serializing.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3d60a452-4df4-42bc-92bf-b2a77f823419-->
<!--TEST RESULTS-->
<!--29e2930c-ba8c-438a-93b2-8a1eda8fa2a8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71889">build#71889</a></summary>


| test_class                         | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71889#01992da8-5650-4c0a-a2a6-ccbaaa9e66cc | FLAKY         | 19/21    | upstream reliability is '98.98477157360406'. current run reliability is '90.47619047619048'. drift is 8.50858 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72018">build#72018</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72018#019934d7-967f-4682-be87-30024d5e0504 | FLAKY         | 20/21    | upstream reliability is '98.70689655172413'. current run reliability is '95.23809523809523'. drift is 3.4688 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72018#019934d7-9680-4eac-9c45-ed5842a3f275 | FLAKY         | 20/21    | upstream reliability is '99.06103286384976'. current run reliability is '95.23809523809523'. drift is 3.82294 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat reviewed (commented):
code all lgtm but i'm not familiar with the semantic nuances

nvartolomei commented on src/v/iceberg/manifest_entry_values.cc:
```diff
@@ -205,19 +225,23 @@ data_file data_file_from_value(struct_value v) {
+    file.value_counts = get_counts_map(std::move(fs[7]), "value_counts");
+    file.null_value_counts = get_counts_map(
+      std::move(fs[8]), "null_value_counts");
+    file.nan_value_counts = get_counts_map(std::move(fs[9]), "distinct_counts");
```
copy paste error? you are assigning twice to `file.nan_value_counts`

nvartolomei replied:
and no test i guess? can we add one?

nvartolomei replied:
note: fix is fixed in the next commit

andrwng replied:
Agh nice catch 🤦 pushed a fix

nvartolomei reviewed (approved):
lgtm

andrwng committed:
iceberg: remove manifest entry distinct_counts

The manifest entry schema used by Redpanda to deserialize manifests
(defined in manifest_entry_type.cc[1]) does not include distinct_counts, as
the field has been removed from the Iceberg spec, but we were previously
trying to still deserialize the distinct_counts field in
manifest_entry_values.cc[2].

This meant that we were deserializing a distinct_counts field with
values of what should have been nan_value_counts. And that we were
deserializing a nan_value_counts fields with what should have been
lower_bounds.

Because lower_bounds (map<int, binary>) and nan_value_counts (map<int,
long>) have different types, this meant that when a manifest included
entries with a non-null lower_bounds fields, Redpanda would fail to
deserialize, returning the following error:

```
ERROR 2025-08-15 13:20:44,446 [shard 1:main] iceberg - metadata_io.h:166 - Exception while parsing iceberg::manifest (path: "redpanda/topic/metadata/52d25526-4b0a-4c2e-b7bf-0c93506d53ae-m4.avro"): std::invalid_argument (Value is not the expected type)
```

or with the recently updated logging:

```
Error parsing 'nan_value_counts' map: Value of 'val' is not the expected type long: actual binary(size_bytes=1)
```

This commit fixes the deserialization error and updates the test
manifest to include a lower and upper bound field that would have caught
this error.

Longer term we should really de/serialize all fields (rather than just
the required fields), and try to figure out a way to make manifest
serialization less error prone. That is left as future work though,
given this bug on its own is a small fix.

[1] https://github.com/redpanda-data/redpanda/blob/dc2a0537ff85dcf048d01eaddfce830ee78c4971/src/v/iceberg/manifest_entry_type.cc#L58-L71
[2] https://github.com/redpanda-data/redpanda/blob/dc2a0537ff85dcf048d01eaddfce830ee78c4971/src/v/iceberg/manifest_entry_values.cc#L219

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x
